src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            int? documentVersion = null;
+            DocumentSnapshot documentSnapshot = null;
+            await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.RazorDocumentUri.GetAbsoluteOrUNCPath(), out documentSnapshot);
+
+                Debug.Assert(documentSnapshot != null, "Failed to get the document snapshot, could not map to document ranges.");
+
+                if (documentSnapshot is null ||
+                    !_documentVersionCache.TryGetDocumentVersion(documentSnapshot, out documentVersion))
+                {
+                    documentVersion = null;
+                }
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = null,
+                    HostDocumentVersion = null
+                };
+            }
+
+            var unmappedDiagnostics = request.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d)).ToArray();
+            if (!filteredDiagnostics.Any())
+            {
+                // No diagnostics left after filtering.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>(),
+                    HostDocumentVersion = documentVersion
+                };
+            }
+
+            return await MapDiagnosticsAsync(
+                request,
+                filteredDiagnostics,
+                documentVersion,
+                documentSnapshot).ConfigureAwait(false);
+
+            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
+            static bool CanDiagnosticBeFiltered(Diagnostic d) =>
+                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
+                 d.Severity != DiagnosticSeverity.Error);
+        }
+
+        private async Task<RazorDiagnosticsResponse> MapDiagnosticsAsync(
+            RazorDiagnosticsParams request,
+            Diagnostic[] filteredDiagnostics,
+            int? documentVersion,
+            DocumentSnapshot documentSnapshot)
+        {
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = filteredDiagnostics,
+                    HostDocumentVersion = documentVersion,
+                };
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
```

Comment:
You only ever use the documentSnapshot for the code document and later on you check if the code document is unsupported on every diagnostic. I'd calculate the code document prior to getting the "filteredDiagnostics" and then do one check for "IsUnsupported" to simplify the code below. It'd also mean this MapDiagnosticsAsync wouldd take a CodeDocument

Reply:
Makes sense, thanks!

Reply:
> and then do one check for "IsUnsupported" to simplify the code below.

Done.

> I'd calculate the code document prior to getting the "filteredDiagnostics" 

Given the above (if we aren't going through these mapping steps for the HTML case), then I believe we should hold off on generating the output document till here as only now do we know we need it.

Reply:
Discussed below, we're holding off on the mapping so I believe this'll need to be kept as is for now.

https://github.com/dotnet/aspnetcore/issues/29007